src/resolve.ts

```diff
@@ -0,0 +1,115 @@
+import { ApiObject } from './api-object';
+import { App } from './app';
+import { Lazy } from './lazy';
+
+/**
+ * Context object for a specific resolution process.
+ */
+export class ResolutionContext {
+
+  /**
+   * The final value of the resolution process. If the resolver invoked
+   * `replaceValue`, this will be the replaced value, otherwise, it is the original
+   * value.
+   */
+  public value: any;
+
+  constructor(
+    /**
+     * Which ApiObject is currently being resolved.
+     */
+    public readonly obj: ApiObject,
+    /**
+     * Which key is currently being resolved.
+     */
+    public readonly key: string[],
+    /**
+     * The value associated to the key currently being resolved.
+     */
+    public readonly originalValue: any) {
+    this.value = originalValue;
+  }
+
+  /**
+   * Replaces the original value in this resolution context
+   * with a new value. The new value is what will end up in the manifest.
+   */
+  public replaceValue(newValue: any) {
+    this.value = newValue;
+  }
+
+}
+
+/**
+ * Contract for resolver objects.
+ */
+export interface IResolver {
+
+  /**
+   * This function is invoked on every property during cdk8s synthesis.
+   * To replace a value, implementations must invoke `context.replaceValue`.
+   */
+  resolve(context: ResolutionContext): void;
+}
+
+/**
+ * Resolvers instanecs of `Lazy`.
+ */
+export class LazyResolver implements IResolver {
+
+  public resolve(context: ResolutionContext): void {
+    if (context.value instanceof Lazy) {
```

Comment:
`instanceof` isn't safe against an `npm link` workflow. Wouldn't you want to do an interface test similar to what you do for `resolve()` ?

Reply:
Not eager to touch this and increase the scope of this PR. Its been like this since forever, so keeping it as is.